### DISCUSSION TITLE
Any plans to support Core3 from m5 stack

### DISCUSSION THREAD

**Discussion by kairos0ne** [Moddable-OpenSource/moddable#1134] [Q&A]
Hey just wanted to check if the core3 from m5stack is in the roadmap for support on your platform ?

**Comment by phoddie** [comment]
It is. We already have support for [a bunch of ESP32-S3 boards](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/esp32.md#platforms-esp32-s3). I think @meganetaaan already has one and may be looking into that.

**Reply by meganetaaan** [reply]
Yes, I am considering adding support for M5Stack Core3 in Moddable.
https://github.com/meganetaaan/stack-chan/issues/177
I have just started working on it and there is no progress yet. I will update as I make progress.

@stc1988 and @washishi also had similar motivations. I would be very happy if they could share their achievements.

**Reply by kairos0ne** [reply]
Amazing news...

**Reply by meganetaaan** [reply]
I have initiated implementation at https://github.com/meganetaaan/moddable/tree/feature/m5stack_cores3.

Here's the current status:
- A new target `esp32/m5stack_cores3` has been created. Use the following command to write the program.

```
mcconfig -m -d -p esp32/m5stack_cores3
```

- I have made it possible to initialize and control the LCD. Many sample codes in the `examples/piu` directory are now functioning.
- Despite errors due to differing vendor IDs (a problem similar to a past [issue](https://github.com/Moddable-OpenSource/moddable/issues/1064) I encountered), the touch control IC is operational.
- No sound is produced from the speaker. While AudioOut runs without any errors, there is no audio output. This could potentially be due to the amp IC not initializing correctly.
- Many peripherals such as cameras and distance sensors have not been implemented yet.

I'm currently in the midst of active development to ensure compatibility with CoreS3. If anyone is interested in assisting, your help would be greatly appreciated.

**Reply by phoddie** [reply]
There's a [PR pending](https://github.com/Moddable-OpenSource/moddable/pull/1165) from @meganetaaan with initial M5Stack CoreS3 support! While it is still a work-in-progress it looks to be impressively far along. If anyone else has a chance to try that out, it would be great to hear about your experience. Thank you!

**Comment by mkellner** [comment]
Looking at https://github.com/m5stack/M5Unified/blob/master/src/utility/Speaker_Class.cpp line 65, I see that bits_per_sample is 16 bit.

More initialization at: https://github.com/m5stack/M5Unified/blob/master/src/M5Unified.cpp#L683

Oh. at line 691: `spk_cfg.i2s_port = I2S_NUM_1;`

In the `m5stack_cores3/manifest.json`, add `"num": 1` as it defaults to 0 in pins/i2s/audio.c.

Maybe try:
```
   "audioOut": {
      "i2s": {
         "bitsPerSample": 16,
         "num": 1,
         ...
```

**Reply by meganetaaan** [reply]
After a number of tests, I've managed to make the sound work!

https://github.com/Moddable-OpenSource/moddable/assets/1918755/dfbf458a-0de5-4fa5-8074-c3e59ed96e34

During the testing process, I discovered a peculiar situation where no sound is produced only when the sample rate is set to 11025. This issue seems to be linked to the way the ESP-IDF handles clock division calculations. When the sample rate is set to a value that doesn't divide evenly, it appears to continuously fluctuate between slightly faster and slower speeds (ranging from 11021 to 11030) to make up for the decimal places. This inconsistency disrupts the synchronization with the AW88298 amp IC's phase-locked loop (PLL).
https://discord.com/channels/1095725099925110847/1115549881508040765/1118192144176386182

To prevent this issue, I've set a default value that's a bit higher than 11025.

On a separate note, it's crucial to specify the sample rate directly on the amp IC. If a sample rate different from the default value is provided in the AudioOut constructor, a way to dynamically adjust the settings is needed. We could potentially implement this by adding a callback to AudioOut, or by creating a class that wraps a M5Stack CoreS3-specific AudioOut.

**Reply by phoddie** [reply]
Congratulations on getting the audio working. Very cool.

Audio hardware comes with so many strange restrictions. The sample rate here seems to be a challenge.

> We could potentially implement this by adding a callback to AudioOut, or by creating a class that wraps a M5Stack CoreS3-specific AudioOut

The latter seems preferable, as it maintains a stable API and the client code doesn't need to be involved at all. You can use the JavaScript return override to do this in a very low-over head way. Something like this:

```js
import AudioOut from "pins/audioout-original";

class CoreS3Audio {
   constructor(options) {
     const a = new AudioOut(options);
     // do whatever you need to for the amp IC using a.sampleRate
     return a;
   }
}
```

**Comment by kairos0ne** [comment]
I’ll test this np, sorry just been busy with work not had much personal time. I’m a complete nube on microcontrollers so not sure I’ll add much value but will defo give it a go.

**Reply by phoddie** [reply]
Much appreciated!
